Unflagging announcement (#434)
nodejs/modules: doc/unflagging-announcement.md

# Announcing core Node.js support for ECMAScript modules

Node.js 13.2.0 ships support for [ECMAScript modules](https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/), known for their `import` and `export` statements. This support was previously behind the `--experimental-modules` flag, which is no longer required; however the implementation remains experimental and subject to change.

## Enabling

> Node.js will treat the following as ES modules when passed to `node` as the initial input, or when referenced by `import` statements within ES module code:

Reviewer: Should we drop "within ES module code" given `import()` can occur in CJS code?

Author: `import()` is an expression, not a statement. “`import` statement” means `import`, not `import()`.

Reviewer 2: So we should add `import()`? Because I don't think we ever treat things differently between `import` and `import()` or do we?

Author: Within a CommonJS context, how is the specifier in `import()` treated? The same as in `require`, or the same as in `import()` in an ESM context?

Author: BTW this was a direct copy from https://nodejs.org/api/esm.html#esm_enabling, so if this needs changing here it should also get updated there.

Reviewer 2: The specifier passed to `import()` is always treated like in `import` statements. This is true for `eval`, CJS, ESM, `new Function`, `vm.createScript`, and all other kinds of context. There's some magic that breaks it for `vm.createSourceTextModule` IIRC but generally it's the same as in the browser: `import` is `import`, there's "nothing" special about references from ESM files. :)

Author: So I think the thinking behind “`import` statements within ES module code” was to remind readers that `import` statements can only exist within ES module code. This is the very top of the docs, so we’re assuming beginner-level knowledge at this point. But maybe I shouldn’t bother, and just assume that people know that `import` is ESM syntax and can’t appear in CommonJS? That’s better than needing to draw a distinction between `import` and `import()`?

* Files ending in `.mjs`.

* Files ending in `.js`, or extensionless files, when the nearest parent `package.json` file contains a top-level field `"type"` with a value of `"module"`.

> * Strings passed in as an argument to `--eval` or `--print`, or piped to `node` via `STDIN`, with the flag `--input-type=module`.

Reviewer 3: probably best not to mention `--print`, since we're about to remove it because of TLA.

Author: Yeek. This was copied straight from https://nodejs.org/api/esm.html#esm_enabling, should that be updated too? Or maybe update that when the change actually happens?

Reviewer 3: the docs should be updated with the actual change. I was thinking it would be best to not advertise it in the announcement.

> Node.js will treat as CommonJS all other forms of input, such as `.js` files where the nearest parent `package.json` file contains no top-level `"type"` field, or string input without the flag `--input-type`. This behavior is to preserve backward compatibility. However, now that Node.js supports both CommonJS and ES modules, it is best to be explicit whenever possible. Node.js will treat the following as CommonJS when passed to `node` as the initial input, or when referenced by `import` statements within ES module code:

Reviewer: Suggest adding a paragraph separator at "Node.js will treat"

* Strings passed in as an argument to `--eval` or piped to `node` via `STDIN`, with the flag `--input-type=commonjs`.

For more information, see [Package Scope and File Extensions](https://nodejs.org/api/esm.html#esm_package_scope_and_file_extensions) and [`--input-type` flag](https://nodejs.org/api/esm.html#esm_code_input_type_code_flag).

## `import` and `export` syntax

Within ES module files and string input, [`import` statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) can reference JavaScript files. Files can be referenced as:

- Relative URLs (`'./file.mjs'`)

- Absolute `file://` URLs (`'file:///opt/app/file.mjs'`)

- Package names (`'es-module-package'`)

- Paths within packages (`'es-module-package/lib/file.mjs'`)

`import` statements that reference ES module files can specify the default export (`import _ from 'es-module-package'`), named exports (`import { shuffle } from 'es-module-package'`) and namespace exports (`import * as fs from 'fs'`). All Node.js built-in packages like `fs` and `path` support all three types of exports.

> `import` statements that reference CommonJS files (all current JavaScript code written for Node.js, [using `require` and `module.exports`](https://nodejs.org/api/modules.html)) can use the CommonJS default export (`import _ from 'commonjs-package'`) only.

Reviewer: Style: Bring forward "only" into "can only use"

`import` statements that reference other types of files such as JSON or Wasm are experimental and require the [`--experimental-json-modules`](https://nodejs.org/api/esm.html#esm_experimental_json_modules) or [`--experimental-wasm-modules`](https://nodejs.org/api/esm.html#esm_experimental_wasm_modules) flags. The non-experimental [`module.createRequire`](https://nodejs.org/api/modules.html#modules_module_createrequire_filename) can be used to import JSON or native addons.

[`export` statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) in ES module files can specify both default and named exports for `import` statements to reference.

> Dynamic `import()` expressions can be used to import ES modules from either CommonJS or ES module files. Note that `import()` returns a promise.

Reviewer: Dynamic import grabs a namespace, not an ES module.

> Dynamic `import()` expressions can be used to import ES namespace objects from either CommonJS or ES module files. Note that `import()` returns a promise for the namespace.

Author: I don’t think most people will know what a namespace is. Is the distinction important?

Reviewer: It's the essence of what the operator does, so I think it's worth explicitly using the term "namespace".

[`import.meta.url`](https://nodejs.org/api/esm.html#esm_import_meta) provides the absolute URL of the current ES module file.

## Files and the `package.json` `"type"` field

Add `"type": "module"` to the `package.json` for your project, and Node.js will treat all `.js` files in your project as ES modules.

If some of your project’s files use CommonJS and you can’t convert your entire project all at once, you can either rename those files to use the `.cjs` extension or put them in a subfolder containing a `package.json` with `{ "type": "commonjs" }`, under which all `.js` files are treated as CommonJS.

> For any file that Node.js tries to load, it will look for a `package.json` in that file’s folder, then that file’s parent folder and so on upwards until it reaches the root of the volume. This is similar to [how Babel searches for `.babelrc` files](https://babeljs.io/docs/en/config-files#file-relative-configuration). This new approach allows Node.js to use `package.json` for package-level metadata and configuration, similar to how it is already used by Babel and other tools.

Reviewer 4: This is either incomplete or I've discovered a bug. The search seems to stop once it reaches a directory named `node_modules`. Additionally, placing a `package.json` with a `"type": "module"` field in a `node_modules` directory will not resolve files with `.js` (or even `.mjs`) extensions as ES modules. Is this the intended behavior?

Author: Please open an issue.

Reviewer 4: I don't know if this is a documentation issue (not educating users of this behavior) or actual unintended behavior. Hence, I don't know which repo to open an issue in. Please clarify and will gladly do so!

We encourage all packages to include the `"type"` field, even if the `"type"` is `"commonjs"`.

## Package entry points and the `package.json` `"exports"` field

There are now two fields that can define entry points for a package: `"main"` and `"exports"`. The `"main"` field is supported in all versions of Node.js, but its capabilities are limited: it only defines the main entry point of the package. A new `package.json` field `"exports"` can also define the main entry point, along with subpaths. It also provides encapsulation, where only the paths explicitly defined in `"exports"` are available for importing. `"exports"` applies to both CommonJS and ES module packages, whether used via `require` or `import`.

> This allows statements like `import 'pkg/feature'` to map to a path like `'./node_modules/pkg/esm/feature.js`. It also causes an error to be thrown for statements like `import 'pkg/esm/feature.js'` unless that path is defined in `"exports"`.

Reviewer 5: Maybe include an example `package.json` snippet here?

Author: Our last announcement didn’t include any code snippets. I was under the impression that blog posts shouldn’t get that detailed. The sections link to the docs, which have examples.

Reviewer 5: > The sections link to the docs, which have examples.

Within this section, there's only links for the experimental "conditional exports" right now. If we don't do code snippets, then please consider adding a link to <https://nodejs.org/api/esm.html#esm_package_entry_points>.

## Common Gotchas

### Mandatory file extensions

A file extension must be provided when using the `import` keyword. Directory indexes (e.g. `'./startup/index.js'`) must also be fully specified.

This behavior matches how `import` behaves in browser environments, assuming a typically configured server.

### No <code>require</code>, <code>exports</code>, <code>module.exports</code>, <code>\_\_filename</code>, <code>\_\_dirname</code>

These CommonJS variables are not available in ES modules.

`require` can be imported into an ES module using [`module.createRequire()`](https://nodejs.org/api/modules.html#modules_module_createrequire_filename).

Equivalents of `__filename` and `__dirname` can be created inside of each file via [`import.meta.url`](https://nodejs.org/api/esm.html#esm_import_meta).

## Writing packages

> At the moment, we recommend writing packages that provide to Node.js only entirely CommonJS or entirely ES module sources. The modules team is working on better support for “dual” packages, that provide CommonJS sources for package consumers using `require` and ES module sources for package consumers using `import`. The current experimental feature covering this use case is [conditional exports](https://nodejs.org/api/esm.html#esm_conditional_exports), and the team hopes to ship that or an alternative by the end of January 2020 if not earlier. See recommended patterns in [Dual CommonJS/ES Module Packages](https://nodejs.org/api/esm.html#esm_dual_commonjses_module_packages).

Reviewer: Maybe this first sentence is less awkward:

> At the moment, we recommend packages are written entirely using CommonJS or entirely using ES module format.

Author: It is, but I don’t want to discourage people continuing to write packages that include ESM sources for use via `"module"`—as in, to provide to build tools, but not to Node.

Reviewer: Retaining the word sources is fine.

> 

Reviewer 6: This announcement has 3 links to the same document yet the document doesn't contain a section on Dual CJS/ESM packages 🤔 

https://nodejs.org/api/esm.html#esm_dual_commonjses_module_packages

Author: That link should be live by the next time the docs are generated, which should happen with the next release: https://github.com/nodejs/node/blob/master/doc/api/esm.md#dual-commonjses-module-packages

# Announcing core Node.js support for ECMAScript modules

Node.js 13.2.0 ships support for [ECMAScript modules](https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/), known for their `import` and `export` statements. This support was previously behind the `--experimental-modules` flag, which is no longer required; however the implementation remains experimental and subject to change.

## Enabling

> Node.js will treat the following as ES modules when passed to `node` as the initial input, or when referenced by `import`:

Reviewer 5: Minor writing tip: in general, avoiding "will" makes things more concrete and is often a good way to simplify prose. ([h/t to Steve Souders](http://blog.stoyanstefanov.com/technical-writing-checklist/))

```suggestion
Node.js now treats the following as ES modules when passed to `node` as the initial input, or when referenced by `import`:
```

* Files ending in `.mjs`.

* Files ending in `.js`, or extensionless files, when the nearest parent `package.json` file contains a top-level field `"type"` with a value of `"module"`.

* Strings passed in as an argument to `--eval` or piped to `node` via `STDIN`, with the flag `--input-type=module`.

Node.js will treat as CommonJS all other forms of input, such as `.js` files where the nearest parent `package.json` file contains no top-level `"type"` field, or string input without the flag `--input-type`. This behavior is to preserve backward compatibility. However, now that Node.js supports both CommonJS and ES modules, it is best to be explicit whenever possible.

> Node.js will treat the following as CommonJS when passed to `node` as the initial input, or when referenced by `import`:

Reviewer 5: ```suggestion
Node.js treats the following as CommonJS when passed to `node` as the initial input, or when referenced by `import`:
```

* Strings passed in as an argument to `--eval` or piped to `node` via `STDIN`, with the flag `--input-type=commonjs`.

For more information, see [Package Scope and File Extensions](https://nodejs.org/api/esm.html#esm_package_scope_and_file_extensions) and [`--input-type` flag](https://nodejs.org/api/esm.html#esm_code_input_type_code_flag).

## `import` and `export` syntax

Within ES module files and string input, [`import` statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) can reference JavaScript files. Files can be referenced as:

- Relative URLs (`'./file.mjs'`)

- Absolute `file://` URLs (`'file:///opt/app/file.mjs'`)

- Package names (`'es-module-package'`)

- Paths within packages (`'es-module-package/lib/file.mjs'`)

`import` statements that reference ES module files can specify the default export (`import _ from 'es-module-package'`), named exports (`import { shuffle } from 'es-module-package'`) and namespace exports (`import * as fs from 'fs'`). All Node.js built-in packages like `fs` and `path` support all three types of exports.

> `import` statements that reference CommonJS files (all current JavaScript code written for Node.js, [using `require` and `module.exports`](https://nodejs.org/api/modules.html)) can only use the CommonJS default export (`import _ from 'commonjs-package'`).

Reviewer 7: But this is only possible for built-in packages until dual cjs/ems support will land, right?

> `import` statements that reference CommonJS files (all current JavaScript code written for Node.js, [using `require` and `module.exports`](https://nodejs.org/api/modules.html)) can only use the CommonJS default export (`import _ from 'commonjs-package'`).

Reviewer 7: ```suggestion
`import` statements that reference built-in packages can only use the CommonJS default export (`import _ from 'commonjs-package'`).
```

Author: No, `import` can import CommonJS.

`import` statements that reference other types of files such as JSON or Wasm are experimental and require the [`--experimental-json-modules`](https://nodejs.org/api/esm.html#esm_experimental_json_modules) or [`--experimental-wasm-modules`](https://nodejs.org/api/esm.html#esm_experimental_wasm_modules) flags. The non-experimental [`module.createRequire`](https://nodejs.org/api/modules.html#modules_module_createrequire_filename) can be used to import JSON or native addons.

[`export` statements](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) in ES module files can specify both default and named exports for `import` statements to reference.

Dynamic `import()` expressions can be used to import ES modules from either CommonJS or ES module files. Note that `import()` returns a promise.

[`import.meta.url`](https://nodejs.org/api/esm.html#esm_import_meta) provides the absolute URL of the current ES module file.

## Files and the `package.json` `"type"` field

> Add `"type": "module"` to the `package.json` for your project, and Node.js will treat all `.js` files in your project as ES modules.

Reviewer 5: ```suggestion
Add `"type": "module"` to the `package.json` for your project, and Node.js treats all `.js` files in your project as ES modules.
```

If some of your project’s files use CommonJS and you can’t convert your entire project all at once, you can either rename those files to use the `.cjs` extension or put them in a subfolder containing a `package.json` with `{ "type": "commonjs" }`, under which all `.js` files are treated as CommonJS.

> For any file that Node.js tries to load, it will look for a `package.json` in that file’s folder, then that file’s parent folder and so on upwards until it reaches the root of the volume. This is similar to [how Babel searches for `.babelrc` files](https://babeljs.io/docs/en/config-files#file-relative-configuration). This new approach allows Node.js to use `package.json` for package-level metadata and configuration, similar to how it is already used by Babel and other tools.

Reviewer 5: ```suggestion
For any file that Node.js tries to load, it looks for a `package.json` in that file’s folder, then that file’s parent folder and so on upwards until it reaches the root of the volume. This is similar to [how Babel searches for `.babelrc` files](https://babeljs.io/docs/en/config-files#file-relative-configuration). This new approach allows Node.js to use `package.json` for package-level metadata and configuration, similar to how it is already used by Babel and other tools.
```
